source/date_time.c

```diff
@@ -0,0 +1,826 @@
+    }
+
+    if (dt->utc_assumed) {
+        dt->timestamp = aws_mk_gm_time(&parsed_time);
+    } else {
+        dt->timestamp = mktime(&parsed_time);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_date_time_to_local_time_str(
+    struct aws_date_time *dt,
+    enum aws_date_format fmt,
+    struct aws_byte_buf *output_buf) {
+    assert(fmt != AWS_DATE_FORMAT_AUTO_DETECT);
+
+    if (AWS_UNLIKELY(fmt == AWS_DATE_FORMAT_AUTO_DETECT)) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    struct tm time;
+    AWS_ZERO_STRUCT(time);
+    aws_local_time(dt->timestamp, &time);
+    char formatted_string[100];
+    AWS_ZERO_ARRAY(formatted_string);
+
+    if (fmt == AWS_DATE_FORMAT_RFC822) {
+        strftime(formatted_string, sizeof(formatted_string), RFC822_DATE_FORMAT_STR_WITH_Z, &time);
+    } else {
+        strftime(formatted_string, sizeof(formatted_string), ISO_8601_LONG_DATE_FORMAT_STR, &time);
+    }
+
+    size_t len = strlen(formatted_string);
+
+    if (output_buf->capacity < len) {
+        return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
+    }
+
+    memcpy(output_buf->buffer, formatted_string, len);
+    output_buf->len = len;
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_date_time_to_utc_time_str(struct aws_date_time *dt, enum aws_date_format fmt, struct aws_byte_buf *output_buf) {
+    assert(fmt != AWS_DATE_FORMAT_AUTO_DETECT);
+
+    if (AWS_UNLIKELY(fmt == AWS_DATE_FORMAT_AUTO_DETECT)) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    struct tm time;
+    AWS_ZERO_STRUCT(time);
+    aws_gm_time(dt->timestamp, &time);
+    char formatted_string[100];
+    AWS_ZERO_ARRAY(formatted_string);
+
+    if (fmt == AWS_DATE_FORMAT_RFC822) {
+        strftime(formatted_string, sizeof(formatted_string), RFC822_DATE_FORMAT_STR_MINUS_Z, &time);
+    } else {
+        strftime(formatted_string, sizeof(formatted_string), ISO_8601_LONG_DATE_FORMAT_STR, &time);
+    }
+
+    size_t len = strlen(formatted_string);
+
+    if (output_buf->capacity < len) {
+        return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
+    }
+
+    memcpy(output_buf->buffer, formatted_string, len);
+    output_buf->len = len;
+
+    return AWS_OP_SUCCESS;
+}
+
+double aws_date_time_as_epoch_secs(struct aws_date_time *dt) {
+    return (double)dt->timestamp;
+}
+
+uint64_t aws_date_time_as_nanos(struct aws_date_time *dt) {
+    return (uint64_t)dt->timestamp * AWS_TIMESTAMP_NANOS;
+}
+
+uint64_t aws_date_time_as_millis(struct aws_date_time *dt) {
+    return (uint64_t)dt->timestamp * AWS_TIMESTAMP_MILLIS;
+}
+
+struct tm s_get_time_struct(struct aws_date_time *dt, bool local_time) {
+    struct tm time;
+    AWS_ZERO_STRUCT(time);
+    if (local_time) {
+        aws_local_time(dt->timestamp, &time);
+    } else {
+        aws_gm_time(dt->timestamp, &time);
+    }
+
+    return time;
+}
+
+uint16_t aws_date_time_year(struct aws_date_time *dt, bool local_time) {
+    struct tm time = s_get_time_struct(dt, local_time);
```

Comment:
So if I want to read out the components of the aws_date_time struct, I have to convert from the timestamp to the gmtime/localtime structure eight times?

Reply:
I suppose we could stuff the tm into the aws_date_time struct.

Reply:
hmm. for this to work we'd need to store two tm, one for utc and one for localtime... not that big a deal though I guess.

Reply:
That seems like a better move, it would do away with some of the other precision issues and structure copying going on.